### DISCUSSION TITLE
How to get rid of INotifyPropertyChanging?

### DISCUSSION THREAD

**Discussion by mikeKuester** [CommunityToolkit/dotnet#637] [Q&A] [answered]
Hi, 
I'm just starting with the MVVM Community Toolkit and looking for a way to get rid of the `INotifyPropertyChanging` implementation.

Lets say I have a large number of instances of a view models and want to force the source generator, just to implement the code for `INotifyPropertyChanged`, but not for `INotifyPropertyChanging`. I know, that the "changing" events will never be used and in order to improve the performance, I don't want all these `OnPropertyChanging` calls.

Instead of inheriting from `ObservableObject`, I tried to use the `INotifyPropertyChangedAttribute` for the class and the generated code looks as expected - but then I got this [MVVMTK0032 MVVM Toolkit warning](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/errors/mvvmtk0032): you don't implement a base type, so you should inherit from  `ObservableObject`

The resulting code is not the same: using the `INotifyPropertyChanged` attribute or inheriting from `ObservableObject`. Therefore I think the warning from the MVVMTK0032 analyzer is not valid.

The warning is ok if I use the `ObservableObject` attribute for a class without a base type, but not for the `INotifyPropertyChanged` attribute.

Is there anything what I'm missing here?

Thank for your work for the toolkit and
best regards,
Mike

**Comment by Sergio0694** [comment, marked as answer]
> "I know, that the "changing" events will never be used and in order to improve the performance, I don't want all these OnPropertyChanging calls."

That is unlikely to make any observable difference, especially if you're using the source generator for your observable property. That's because the generator will also generate and cache all the `PropertyChangingEventArgs` for each property, so the `OnPropertyChanging` call will simply do nothing. As in, it will see the event is `null`, and just not even raise it at all.

On the other hand, using `ObservableObject` is recommended because it allows you to reuse the same base code across all your viewmodels, whereas if you used `[INotifyPropertyChanged]` in each of them, you'd end up duplicating a lot of code and increasing your binary size unnecessarily.

I would recommend just inheriting from `ObservableObject` and using `[ObservableProperty]`.

If you really want to, you can also set `"MVVMTOOLKIT_DISABLE_INOTIFYPROPERTYCHANGING"` in your app context switches. This will make the whole feature be completely disabled, so that even direct calls from other places will just not do anything 🙂

**Reply by mikeKuester** [reply]
Thanks for your fast answer!

Ok, if its not making an observable difference, that sounds good. And you're right - the caching of the `PropertyChangingEventArgs` is the main key here.
